File: kevinb9n/src/test/java/site/kevinb9n/math/DoubleSubject.java
package site.kevinb9n.math;

import static java.lang.Math.abs;

import com.google.common.base.Preconditions;
import com.google.common.truth.ComparableSubject;
import com.google.common.truth.Fact;
import com.google.common.truth.FailureMetadata;
import org.checkerframework.checker.nullness.compatqual.NullableDecl;

public final class DoubleSubject extends ComparableSubject<Double> {
  private static final long NEG_ZERO_BITS = Double.doubleToLongBits(-0.0D);
  private final Double actual;

  final Fact butWas() {
    return Fact.fact("but was", actualCustomStringRepresentation());
  }

  DoubleSubject(FailureMetadata metadata, @NullableDecl Double actual) {
    super(metadata, actual);
    this.actual = actual;
  }

  public DoubleSubject.TolerantDoubleComparison isWithin(final double tolerance) {
    return new DoubleSubject.TolerantDoubleComparison() {
      public void of(double expected) {
        Double actual = DoubleSubject.this.actual;
        Preconditions.checkNotNull(actual, "actual value cannot be null. tolerance=%s expected=%s", tolerance, expected);
        DoubleSubject.checkTolerance(tolerance);
        if (!(abs(actual - expected) <= abs(tolerance))) {
          DoubleSubject.this.failWithoutActual(
            Fact.fact("expected", String.valueOf(expected)),
            DoubleSubject.this.butWas(), Fact.fact("outside tolerance", String.valueOf(tolerance)));
        }
      }
    };
  }

  public DoubleSubject.TolerantDoubleComparison isNotWithin(final double tolerance) {
    return new DoubleSubject.TolerantDoubleComparison() {
      public void of(double expected) {
        Double actual = DoubleSubject.this.actual;
        Preconditions.checkNotNull(actual, "actual value cannot be null. tolerance=%s expected=%s", tolerance, expected);
        DoubleSubject.checkTolerance(tolerance);
        if (!(abs(actual - expected) > abs(tolerance))) {
          DoubleSubject.this.failWithoutActual(Fact.fact("expected not to be",
            String.valueOf(expected)),
            DoubleSubject.this.butWas(), Fact.fact("within tolerance", String.valueOf(tolerance)));
        }

      }
    };
  }

  static void checkTolerance(double tolerance) {
    Preconditions.checkArgument(!Double.isNaN(tolerance), "tolerance cannot be NaN");
    Preconditions.checkArgument(tolerance >= 0.0D, "tolerance (%s) cannot be negative", tolerance);
    Preconditions.checkArgument(Double.doubleToLongBits(tolerance) != NEG_ZERO_BITS, "tolerance (%s) cannot be negative", tolerance);
    Preconditions.checkArgument(tolerance != 1.0D / 0.0, "tolerance cannot be POSITIVE_INFINITY");
  }

  public abstract static class TolerantDoubleComparison {
    public abstract void of(double var1);

    /** @deprecated */
    @Deprecated
    public boolean equals(@NullableDecl Object o) {
      throw new UnsupportedOperationException("If you meant to compare doubles, use .of(double) instead.");
    }

    /** @deprecated */
    @Deprecated
    public int hashCode() {
      throw new UnsupportedOperationException("Subject.hashCode() is not supported.");
    }
  }
}

File: kevinb9n/src/test/java/site/kevinb9n/math/DoubleSubject.kt
package site.kevinb9n.math

import com.google.common.base.Preconditions
import com.google.common.truth.ComparableSubject
import com.google.common.truth.Fact
import com.google.common.truth.FailureMetadata
import org.checkerframework.checker.nullness.compatqual.NullableDecl

class DoubleSubject(metadata: FailureMetadata, val actual: Double?) :
  ComparableSubject<Double>(metadata, actual) {
  fun isWithin(tolerance: Double): TolerantDoubleComparison {
    return object : TolerantDoubleComparison() {
      override fun of(expected: Double) {
        val actual = this@DoubleSubject.actual!!
        checkTolerance(tolerance)
        if (Math.abs(actual - expected) > Math.abs(tolerance)) {
          this@DoubleSubject.failWithoutActual(
            Fact.fact("expected", expected.toString()),
            butWas(), Fact.fact("outside tolerance", tolerance.toString()))
        }
      }
    }
  }

  fun isNotWithin(tolerance: Double): TolerantDoubleComparison {
    return object : TolerantDoubleComparison() {
      override fun of(expected: Double) {
        val actual = this@DoubleSubject.actual!!
        checkTolerance(tolerance)
        if (Math.abs(actual - expected) <= Math.abs(tolerance)) {
          this@DoubleSubject.failWithoutActual(Fact.fact("expected not to be", expected.toString()),
            butWas(), Fact.fact("within tolerance", tolerance.toString()))
        }
      }
    }
  }

  abstract class TolerantDoubleComparison {
    abstract fun of(expected: Double)
    override fun equals(other: Any?): Boolean {
      throw UnsupportedOperationException("If you meant to compare doubles, use .of(double) instead.")
    }

    override fun hashCode(): Int {
      throw UnsupportedOperationException("Subject.hashCode() is not supported.")
    }
  }

  fun butWas(): Fact {
    return Fact.fact("but was", actualCustomStringRepresentation())
  }

  companion object {
    private val NEG_ZERO_BITS = java.lang.Double.doubleToLongBits(-0.0)
    fun checkTolerance(tolerance: Double) {
      Preconditions.checkArgument(!java.lang.Double.isNaN(tolerance), "tolerance cannot be NaN")
      Preconditions.checkArgument(tolerance >= 0.0, "tolerance (%s) cannot be negative", tolerance)
      Preconditions.checkArgument(java.lang.Double.doubleToLongBits(tolerance) != NEG_ZERO_BITS, "tolerance (%s) cannot be negative", tolerance)
      @Suppress("DIVISION_BY_ZERO")
      Preconditions.checkArgument(tolerance != 1.0 / 0.0, "tolerance cannot be POSITIVE_INFINITY")
    }
  }
}
